Servers/DHCPClient/DHCPv4Client.cpp

```diff
@@ -0,0 +1,279 @@
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "DHCPv4Client.h"
+#include <AK/ByteBuffer.h>
+#include <AK/Function.h>
+#include <LibCore/SocketAddress.h>
+#include <LibCore/Timer.h>
+#include <stdio.h>
+
+//#define DHCPV4CLIENT_DEBUG
+
+static void send(const InterfaceDescriptor& iface, const DHCPv4Packet& packet, Core::Object*)
+{
+    int fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_UDP);
+    if (fd < 0) {
+        dbg() << "ERROR: socket";
+        return;
+    }
+
+    if (setsockopt(fd, SOL_SOCKET, SO_BINDTODEVICE, iface.m_ifname.characters(), IFNAMSIZ) < 0) {
+        dbg() << "ERROR from setsockopt(SO_BINDTODEVICE): " << strerror(errno);
+        return;
+    }
+
+    sockaddr_in dst;
+    memset(&dst, 0, sizeof(dst));
+    dst.sin_family = AF_INET;
+    dst.sin_port = htons(67);
+    dst.sin_addr.s_addr = IPv4Address { 255, 255, 255, 255 }.to_u32();
+    memset(&dst.sin_zero, 0, sizeof(dst.sin_zero));
+
+    auto rc = sendto(fd, &packet, sizeof(packet), 0, (sockaddr*)&dst, sizeof(dst));
+    if (rc < 0) {
+        dbg() << "sendto failed with " << strerror(errno);
+        // FIXME: what do we do here?
+    }
+}
+
+static void set_params(const InterfaceDescriptor& iface, const IPv4Address& ipv4_addr, const IPv4Address& netmask, const IPv4Address& gateway)
+{
+    int fd = socket(AF_INET, SOCK_DGRAM, IPPROTO_IP);
+    if (fd < 0) {
+        dbg() << "ERROR: socket";
+        return;
+    }
+
+    struct ifreq ifr;
+    memset(&ifr, 0, sizeof(ifr));
+    strncpy(ifr.ifr_name, iface.m_ifname.characters(), IFNAMSIZ);
+
+    // set the IP address
+    ifr.ifr_addr.sa_family = AF_INET;
+    ((sockaddr_in&)ifr.ifr_addr).sin_addr.s_addr = ipv4_addr.to_in_addr_t();
+
+    if (ioctl(fd, SIOCSIFADDR, &ifr) < 0) {
+        dbg() << "ERROR: ioctl(SIOCSIFADDR)";
+    }
+
+    // set the network mask
+    ((sockaddr_in&)ifr.ifr_netmask).sin_addr.s_addr = netmask.to_in_addr_t();
+
+    if (ioctl(fd, SIOCSIFNETMASK, &ifr) < 0) {
+        dbg() << "ERROR: ioctl(SIOCSIFADDR)";
+    }
+
+    // set the default gateway
+    struct rtentry rt;
+    memset(&rt, 0, sizeof(rt));
+
+    rt.rt_dev = const_cast<char*>(iface.m_ifname.characters());
+    rt.rt_gateway.sa_family = AF_INET;
+    ((sockaddr_in&)rt.rt_gateway).sin_addr.s_addr = gateway.to_in_addr_t();
+    rt.rt_flags = RTF_UP | RTF_GATEWAY;
+
+    if (ioctl(fd, SIOCADDRT, &rt) < 0) {
+        dbg() << "Error: ioctl(SIOCADDRT)";
+    }
+}
+
+DHCPv4Client::DHCPv4Client(Vector<InterfaceDescriptor> ifnames)
+    : m_ifnames(ifnames)
+{
+    m_server = Core::UDPServer::construct(this);
+    m_server->on_ready_to_receive = [this] {
+        auto buffer = m_server->receive(sizeof(DHCPv4Packet));
+        dbg() << "Received " << buffer.size() << " bytes";
+        auto& packet = *(DHCPv4Packet*)buffer.data();
+        process_incoming(packet);
```

Comment:
We're forgetting to validate the incoming packet size here.

Reply:
If only I had the foresight to do all of these sorts of checks in the first pass...I'd be one happy developer :^)